src/dstack/_internal/core/services/ssh/tunnel.py

```diff
@@ -161,22 +182,24 @@ def proxy_command(self) -> Optional[List[str]]:
         ]
 
     def open(self) -> None:
-        # Using stderr=subprocess.PIPE may block subprocess.run.
-        # Redirect stderr to file to get ssh error message
-        with tempfile.NamedTemporaryFile(delete=False) as f:
-            try:
-                r = subprocess.run(
-                    self.open_command(), stdout=subprocess.DEVNULL, stderr=f, timeout=SSH_TIMEOUT
-                )
-            except subprocess.TimeoutExpired as e:
-                msg = f"SSH tunnel to {self.destination} did not open in {SSH_TIMEOUT} seconds"
-                logger.debug(msg)
-                raise SSHError(msg) from e
-        with open(f.name, "r+b") as f:
-            error = f.read()
-        os.remove(f.name)
+        # We cannot use `stderr=subprocess.PIPE` here since the forked process (daemon) does not
+        # close standard streams (it calls `daemon(3)` with `noclose` set to `1`), therefore
+        # we will wait EOF from the pipe as long as the daemon exists.
+        try:
+            r = subprocess.run(
+                self.open_command(),
+                stdout=subprocess.DEVNULL,
+                stderr=subprocess.DEVNULL,
+                timeout=SSH_TIMEOUT,
+            )
+        except subprocess.TimeoutExpired as e:
+            msg = f"SSH tunnel to {self.destination} did not open in {SSH_TIMEOUT} seconds"
+            logger.debug(msg)
+            raise SSHError(msg) from e
         if r.returncode == 0:
             return
+        with open(self.log_path, "rb") as f:
+            error = f.read()
```

Comment:
If the tunnel is restarted multiple times (e.g. [here](https://github.com/dstackai/dstack/blob/7245280f71f9045655c8f35a3098d6cc12af702b/src/dstack/_internal/server/services/gateways/connection.py#L72)), won't `self.log_path` accumulate error messages from all restarts and lead to misleading or unintelligible logs and exceptions?

Reply:
**upd**: the example is using `SSHTunnel.aopen`, so it won't be the case there. But then another question is — will `SSHTunnel.aopen` detect any errors considering it is not using `self.log_path`?

Reply:
Good catch, thanks!

1. I've added a cleanup routine to `open`/`aopen` methods.
2. After some digging, I've found conditions when ssh blocks stderr (i.e. does not close stderr stream):
    * when `-v` CLI flag is used (we don't use it)
    * when `ProxyJump` is used (we use it, at least with the sync interface, see `SSHAttach`)
    
    the async interface () is also blocked under either of these conditions, so I used the same logic with the log file: https://github.com/dstackai/dstack/pull/1675/commits/1a9fc3be7cb84e25baa3bdf8c7d0095c770124bf#diff-aa9ac39757de67663ddb3e073012670ae11d79d80fbec53a9e6f823bd5225d47R216
3. `subprocess.run` [kills the process on timeout](https://github.com/python/cpython/blob/ff3bc82f7c9882c27aad597aac79355da7257186/Lib/subprocess.py#L551-L552), added `proc.kill()` to `aopen()`: https://github.com/dstackai/dstack/pull/1675/commits/1a9fc3be7cb84e25baa3bdf8c7d0095c770124bf#diff-aa9ac39757de67663ddb3e073012670ae11d79d80fbec53a9e6f823bd5225d47R210